app/src/main/java/com/babylon/wallet/android/domain/usecases/RecoverAccountsForFactorSourceUseCase.kt

```diff
@@ -0,0 +1,208 @@
+import rdx.works.profile.data.model.pernetwork.usedAccountDerivationIndices
+import rdx.works.profile.derivation.model.KeyType
+import rdx.works.profile.derivation.model.NetworkId
+import rdx.works.profile.di.coroutines.IoDispatcher
+import rdx.works.profile.domain.GetProfileUseCase
+import timber.log.Timber
+import javax.inject.Inject
+
+class RecoverAccountsForFactorSourceUseCase @Inject constructor(
+    private val resolveAccountsLedgerStateUseCase: ResolveAccountsLedgerStateUseCase,
+    private val getProfileUseCase: GetProfileUseCase,
+    private val ledgerMessenger: LedgerMessenger,
+    @IoDispatcher private val defaultDispatcher: CoroutineDispatcher
+) {
+
+    private var nextDerivationPathOffset: Int = 0
+
+    private val _interactionState = MutableStateFlow<InteractionState?>(null)
+    val interactionState: Flow<InteractionState?> = _interactionState.asSharedFlow()
+
+    fun reset() {
+        nextDerivationPathOffset = 0
+    }
+
+    suspend operator fun invoke(
+        recoveryFS: RecoveryFactorSource,
+        currentNetworkId: NetworkId? = null
+    ): Result<List<AccountWithOnLedgerStatus>> {
+        return withContext(defaultDispatcher) {
+            _interactionState.update { null }
+            val profile = if (getProfileUseCase.isInitialized()) getProfileUseCase.invoke().firstOrNull() else null
+            val networkId = currentNetworkId ?: Radix.Gateway.mainnet.network.networkId()
+            val indicesToScan =
+                computeIndicesToScan(
+                    derivationPathScheme = recoveryFS.derivationPathScheme,
+                    networkId = networkId,
+                    factorSourceID = recoveryFS.factorSourceId,
+                    profile = profile
+                )
+            val derivedAccounts = when (recoveryFS) {
+                is RecoveryFactorSource.Device -> {
+                    deriveAccountsForDeviceFactorSource(recoveryFS, indicesToScan, networkId)
+                }
+
+                is RecoveryFactorSource.Ledger -> {
+                    val derivedAccountsResult = derivedAccountsForLedger(recoveryFS, indicesToScan, networkId)
+                    if (derivedAccountsResult.isSuccess) {
+                        derivedAccountsResult.getOrNull().orEmpty()
+                    } else {
+                        _interactionState.update { null }
+                        return@withContext Result.failure(
+                            derivedAccountsResult.exceptionOrNull() ?: Exception("Failed to derive public keys with ledger")
+                        )
+                    }
+                }
+
+                is RecoveryFactorSource.VirtualDeviceFactorSource -> {
+                    deriveAccountsForVirtualDeviceFactorSource(recoveryFS, indicesToScan, networkId)
+                }
+            }
+            val resolvedAccounts = resolveAccountsLedgerStateUseCase(derivedAccounts)
+            if (isActive) {
```

Comment:
Why do we check if couroutine is active here? Is it needed ?

Reply:
leftover from initial implementation. Will remove